### ISSUE TITLE
unit tests

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#1805]
I would like to open a discussion (before I work on a proof of concept) to introduce proper unit tests into our testsuite.

We are currently not making the distinction between unit tests (for example for IndexSet) and integration tests (that assemble/solve a complete finite element problem) and functional tests (let's dump out shape functions in quadrature points to make sure the output won't change in the future).

A distinction between these tests would allow us to use a unit testing framework for the unit tests while keeping the current infrastructure for the other tests. A distinction has several advantages:
- A unittesting framework would run unittests much faster by avoiding configuring, compiling, linking, and executing too many binaries.
- While we probably won't find the time to change all old tests, a system in place would at least allow us to classify new tests correctly.
- A split in tests gives us a natural division between "fast to run" and "slow to run". Imagine being able to run the testsuite after every compile and actually do test-driven design.

Questions:
- Do you have any good arguments against this approach?
- Preference for a testing framework? I have used ![catch](https://github.com/philsquared/Catch) in the past and I love the simplicity, but I am open for suggestions.
- Integration with ctest/cdash: I assume we would create a directory unittests/ that generates a single ctest test (by compiling unittests/*cc into a single executable).
- I don't know of a framework that supports checking and comparing bulk output with tolerance in the numdiff style. In catch you can use ```REQUIRE( performComputation() == Approx( 2.1 ) );``` that compares with a floating point tolerance, but that doesn't scale well to say checking a 10x10 matrix. We could a) not do this or b) figure out a way to catch the output printed to ```deallog``` and compare to a reference string/file specified somehow (optionally of course). I am not quite sure...

**Comment by bangerth** [comment]
Do you have an idea what fraction of tests we have that actually are unit tests, i.e., that do all of the correctness checking internally? These are presumably a superset of those tests that simply has `OK` in the output files. They should be relatively simple to convert.

Is the only thing you want to address the slowness of the testsuite? I usually do testdriven design, but then I probably also have more horsepower than most others...

**Comment by Rombur** [comment]
I think that's a good idea. I have used catch and boost for unit testing. They have very similar capabilities. 
I don't think that they would be a problem to check the values of matrices. You can just save the reference matrices in files and read them when you run the test. Boost has also a function ```per_element``` to compare containers that we may be able to use.

**Comment by tjhei** [comment]
> Do you have an idea what fraction of tests we have that actually are unit tests, i.e., that do all of the correctness checking internally?

Crude estimate (output files with less than 3 lines):
```
$ wc -l */*output | awk '$1 < 3' | wc -l
632
$ ll */*output | wc -l
3524
```
> Is the only thing you want to address the slowness of the testsuite?

Reasons that come to mind:
- faster
- better error output (clickable file/line in IDE)
- more modular tests (so we will have more tests that are simpler instead of a large indexset_01.cc that checks 10 different things)
- gives natural splitting between fast/slow tests
- more concise tests (less boilerplate)

**Comment by bangerth** [comment]
> Crude estimate (output files with less than 3 lines):
 > [...]

That's a sizable fraction.

>   * more modular tests (so we will have more tests that are simpler
>     instead of a large indexset_01.cc that checks 10 different things)

Yes, these are historical accidents. I don't think we would write tests 
this way any more today. Maybe someone needs to hire a student worker 
once to break these tests into their individual components.

I'm not super excited about having *two* ways to write tests, and given 
the number of tests we currently have, I can't see a way to completely 
convert things either. Although I imagine something of the following 
kind could work:
* Create a class that instead of writing output to files, stores it in
   an ostreamstring buffer
* Upon destruction compare the contents of the buffer with the contents
   of the 'output' file.
* Abort if fails, succeed otherwise.

If we're clever, we could probably do this in a way that would require 
only mechanical replacement of `deallog` in all tests, and other 
scriptable jobs. I imagine one could convert a large number of tests 
this way to do the comparison internally, rather than externally.


Do any of these test frameworks have provisions to run tests on multiple 
threads on parallel computers? Do they keep track of the run time of 
each test so they can sort them for the next parallel run? These are 
things I really like about ctest.

**Comment by Rombur** [comment]
@bangerth As far as I know they don't but it's possible to use ctest and catch together.

**Comment by tjhei** [comment]
> * Create a class that instead of writing output to files, stores it in
>    an ostreamstring buffer
> * Upon destruction compare the contents of the buffer with the contents
>   of the 'output' file.

Yes, that is similar to what I thought. If you want to use numdiff you need to write it to a file, though. I don't know if a library version of something like numdiff exists.

> Do any of these test frameworks have provisions to run tests on multiple
> threads on parallel computers?

Catch and googletest do not (see [FAQ for googletest](https://code.google.com/p/googletest/wiki/FAQ#Why_don't_Google_Test_run_the_tests_in_different_threads_to))
I am not sure how important that is because some individual tests will use multiple threads anyways.

This leads to another question: what about MPI based tests? We might want to keep those as is for now.

**Comment by Rombur** [comment]
I have used boost and catch with mpi without any problem.

**Comment by tjhei** [comment]
> I have used boost and catch with mpi without any problem.

And that works with >1 rank? Then you would need to run the testsuite with mpirun...

**Comment by Rombur** [comment]
> And that works with >1 rank? Then you would need to run the testsuite with mpirun...

Yes, it works with several cores and yes, you need to use mpirun. I usually separate the testsuite in one serial and one for mpi.

**Comment by bangerth** [comment]
If you have 64-core machines, the ability to run test on multiple threads becomes really important to you to increase throughput. I can currently run the debug versions of all tests in ~20 minutes and that allows me to do it as part of regular development. I don't want us to go to a solution that runs one test at a time even on a 64-core machine.

**Comment by bangerth** [comment]
> Yes, that is similar to what I thought. If you want to use numdiff you
> need to write it to a file, though. I don't know if a library version of
> something like numdiff exists.

We can always run numdiff by calling `system(...)` from the executable 
itself (e.g., from the destructor of our special log class).

**Comment by tjhei** [comment]
Turns out (googletest supports sharding)[https://code.google.com/p/googletest/wiki/AdvancedGuide#Distributing_Test_Functions_to_Multiple_Machines] which we might be able to abuse. This would run multiple processes instead of multithreaded, which is probably better anyways because it won't interfere with individual tests using multiple tests.

**Comment by bangerth** [comment]
But sharding doesn't balance the workload of tests because it seems to statically partition tests to individual processors. ctest does dynamic assignment, based on how long a test previously ran: it starts many threads and runs the longest tests first.

**Comment by Rombur** [comment]
But you can use the unit tests framework to create the tests and then, use ctest to run the tests individually. Here is a code that just does that: https://github.com/Fraser999/MaidSafe/blob/846ce37c23bd86ee261e439919dfd7000a8f9372/cmake_modules/add_catch_tests.cmake

**Comment by tjhei** [comment]
It discovers tests by parsing the source code? Ugh. But the idea is interesting. I think we could parse the output of ```./test_executable -l``` and create separate ctest tests from it. Neat.

**Comment by tjhei** [comment]
Here is what I have come up with: https://github.com/tjhei/catch_ctest_demo
Works perfectly!

**Comment by bangerth** [comment]
> But you can use the unit tests framework to create the tests and then, use
> ctest to run the tests individually.

But in that case, what do we gain compared to what we have now?

**Comment by tjhei** [comment]
> But in that case, what do we gain compared to what we have now?

I think all advantages still hold. You still have a single binary that gets compiled/linked once but gets executed repeatedly with different command line arguments.

**Comment by bangerth** [comment]
I don't really see the advantage of that. Say, a test fails by producing different output than expected. How will I know how to execute it in a debugger? How will I know which function to step into to see what is going on? We currently have ~400,000 lines of .cc files in the `tests/` directory -- that's going to yield one rather unwieldy executable to debug, even if you break it into a number of pieces.

**Comment by tjhei** [comment]
I see your point and I am also not totally convinced how much of a win all this is, but running individual tests and debugging them is easy:
```
$ ctest -j8
[...]
The following tests FAILED:
	  2 - some_failing_test (Failed)
```
Ctest tests can not have spaces so I replace them by underscores for now). ```catch_test``` is the name of the unittest executable. If you run it with a test name, it will just execute that test:
```
$ ./catch_test "some failing test"
catch_test is a Catch v1.2.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
some failing test
-------------------------------------------------------------------------------
/users/heister/Dropbox/work/deal.II/catch_ctest/main.cc:16
...............................................................................

/users/heister/Dropbox/work/deal.II/catch_ctest/main.cc:18: FAILED:
  REQUIRE( 1==1+1 )
with expansion:
  1 == 2

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed
```
Of course we can use gdb too:
```
$ gdb --args ./catch_test "some failing test"
(gdb) catch throw
Catchpoint 1 (throw)
(gdb) r
Starting program: /home/heister/Dropbox/work/deal.II/catch_ctest/catch_test some\ failing\ test
[...]
Catchpoint 1 (exception thrown), 0x00007ffff7b348b0 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) frame 2
#2  0x0000000000473c9f in ____C_A_T_C_H____T_E_S_T____16 () at /home/heister/Dropbox/work/deal.II/catch_ctest/main.cc:18
18	  REQUIRE(1==1+1);
```

**Comment by bangerth** [comment]
I understand that it is *possible*, but more complicated than it was before (especially if the failure is not simply a thrown exception but just wrong output).

So the question I keep coming back to is this: If we go this route, what do we gain
* in terms of things we couldn't do before
* in terms of things that were clumsy or complicated before
* in terms of speed of running the testsuite?

From the arguments above, we already know that
* we will have to have two different ways of writing tests
* they will be more awkward to debug than before
* compiling each test may potentially take a long time if throw lots of existing tests together
* they will not as easily integrate into parallel test runs as ctest currently does..

**Comment by tjhei** [comment]
> I understand that it is possible, but more complicated than it was before (especially if the failure is not simply a thrown exception but just wrong output).

Yes, you are raising good points. I am not convinced either, but that is why I opened this discussion. :-)

> If we go this route, what do we gain

I assume we will gain a lot of speed (configuring and linking one instead of 500+ executables), but this should be tested.

> they will not as easily integrate into parallel test runs as ctest currently does..

With a trick similar to what I wrote above, they will integrate just fine with the current ctest architecture and run in parallel.

> we will have to have two different ways of writing tests

I agree. This is probably the main drawback.

**Comment by bangerth** [comment]
So I tried this out: I collated all 294 tests in `base/` into a single `.cc` file of some 32,000 lines with this kind of script:
```
for i in [a-w]*cc ; do base=`echo $i | sed s/\.cc//` ; echo "namespace $base {" ; cat $i | perl -pi -e "s/(#include.*)/}\n\1\nnamespace $base {/g;" | perl -pi -e "s/int main/void foo_$base/g;" ; echo "}" ; done > x.cc
```
With some minor editing one can actually compile this. Compiling this one file takes around 1m13s -- quite a long time, in fact, compared to compiling a single test: for example, `is_finite_complex.cc` takes about 2 seconds to compile. So as far as turnaround for individual tests one may want to work with, the one big file approach is definitely not faster.

Of course, running this one test will also take a very long time, so to determine whether your latest fix helped the problem will also take quite some time unless you recall the exact syntax to run tests. (I.e., the equivalent of, for example, `ctest -R fe_values_views` which captures a whole bunch of tests.)

**Comment by tjhei** [comment]
Interesting! 

> So as far as turnaround for individual tests one may want to work with, the one big file approach is definitely not faster.

I think we would still have separate .cc files (for example one .cc for all complex tests). Those .cc's would be linked into a single executable. This way you typically only need to recompile that one .cc and then link (the link will be a more expensive than currently of course). No need to recompile every single test.

> will also take quite some time unless you recall the exact syntax to run tests.

Yes, you will be able to do ```ctest -R fe_values_views``` just like before to run specific unittests.

If we decide that it might be worth it to look into (if we can accept that we will have two different approaches to write tests), we should test how much faster running all tests and compiling/running an individual test is.

**Comment by bangerth** [comment]
Here are some timings:

* With library already built, re-compiling all tests and running the test suite, `ctest -j16` reports 4970 seconds
* Just touching the libraries (so re-linking is necessary, but not re-compiling any of the tests): 2025 seconds
* Touching all test executables via `for i in tests/*/*.{debug,release} ; do x=`echo $i | perl -pi -e 's#.*/##;'` ; touch $i/$x ; done`, then running the tests again (i.e., requiring no re-linking, only executing tests): 1500 seconds

In other words, what seems to me is that we have:
* Running tests: 1500 seconds
* Linking tests: 525 seconds
* Compiling tests: 2900 seconds

**Comment by tamiko** [comment]
@bangerth @tjhei I'm curious how much time we will save for compiling (and linking) with combined test drivers.

**Comment by tjhei** [comment]
I will try to prepare a comparison.

**Comment by tjhei** [comment]
I created 500 unittests that run very fast (they just check Tensor::unroll) in various different configurations. Here is the repo: https://github.com/tjhei/dealii/tree/unit_test_testing

Lessons:
- even running 500 executables (or rerunning the same with a different parameter) takes significant time
- we can save a lot of build&link time, but we still pay for the runtime cost
- It would have been more realistic if I had picked a test that takes some time to run.

Conclusion: Just converting the tests into unittests only saves link time and (maybe) compile time (including all .cc into a single file needs some modifications). Nevertheless, having a set of unittests (even only for new functionality) that run in a few seconds would be amazing. If we put a new framework in place now, we can start writing new faster tests that we can easily run in CI.

Detailed notes:
```
*** 500 tests old style

Old-style: 500x .cc, 500x .output. Even though the test runs very fast, we pay
for 500x compile&link&run.

cd unit-old; . ./setup.sh

$ cd $DEAL_II_BUILD; time ninja setup_tests_unit-old
real	0m10.130s
$ time ctest -R "unit-old" -j 8
real	6m31.815s
$ touch test_363.debug.output && time ctest -R "unit-old" -j 8
real	1m44.967s
$ touch test_363.debug.output && time ctest -R "unit-old/test_363" -j 8
real	0m2.260s

*** 500 tests in a single .cc compiled in one exe

This is the optimal case without linker/compile/runtime overhead.

cd unit-single; ./setup.sh

$ time cmake -D DEAL_II_DIR=/ssd/deal-builds/installed-dev/ .
real	0m0.362s
$ time make -j 8
real	0m14.779s
$ time ./catcher
real	0m0.279s
$ time 

*** 500 tests in 500 separate .ccs compiled in one exe

Compiling each test separately takes a long time, but we only have to link
once. So we are 2x faster than the current approach. Advantage: we only have
to compile stuff that has changed and we only link once instead of 500 times.

cd unit-split; . ./setup.sh

$ time cmake -D DEAL_II_DIR=/ssd/deal-builds/installed-dev/ .
real	0m1.247s
$ time make -j 8
real	3m28.562s
$ time ./catcher
real	0m0.297s
$ touch test_42.cc && time make && ./catcher
real	0m13.969s


** 500 tests in 500 separate .ccs, included together, using catch->ctest  (unit-inc)

This will generate separate ctest tests for each unittest. Running the tests
will execute the binary 500 times, which slows down executing. Advantage:
Usage is more or less identical to the current approach and only the tests you
are interested in are run.

cd unit-inc; . ./setup.sh

$ time cmake -D DEAL_II_DIR=/ssd/deal-builds/installed-dev/ .
real	0m1.098s
$ time make 
real	0m21.437s
$ time make setup_tests
real	0m0.703s
$ time ctest -j 8
real	0m20.478s
```

**Comment by bangerth** [comment]
We could relatively easily write a script that renamed the `main()` function in each test to `main_$testname`, add a file `all.cc` that simply `#includes` all other `.cc` files in a directory, and compile that once. The `all.cc`'s `main()` function would then just call all other files' renamed `main` function. One would need a bit of glue code to reset the `deallog` output file in between. It may be interesting to see how long it takes to compile something of this sort, but I suspect not terribly long.

My problem with the solution of throwing everything into one big file is that on any given platform, we always have ~20 tests that fail. This makes debugging difficult because you can't just run a test under a debugger and wait till it aborts somewhere because the abort may in fact be in an unrelated test earlier on. The approach we have right now, with one test per file, makes it surpremely easy to debug because one knows exactly which executable to debug, where to set breakpoints, etc. I would like every solution we can come up with to retain this feature.

**Comment by bangerth** [comment]
I think we should discuss what to do with this at the upcoming hackathon. I'm not convinced that our current solution is particularly bad -- I like the fact that we have lots of tests and that I can slice and dice them whichever way is necessary for the development I'm currently doing. I also think that debugging individual programs is far easier (and easier to teach!) than having something in one very large executable.

**Comment by simonsticko** [comment]
Has there been any more discussion on this topic? Or a decision?

One could argue that using a unit testing framework would make it easier for new contributors to write good unit tests.

**Comment by masterleinad** [comment]
> One could argue that using a unit testing framework would make it easier for new contributors to write good unit tests.

I don't quite agree. You will still need to learn the unit testing framework and I believe the one we are using right now is simple enough. The only advantage I see is running the test suite in a shorter amount of time.

**Comment by tjhei** [comment]
> The only advantage I see is running the test suite in a shorter amount of time.

That and having to write less code. This is what we do in ASPECT:
https://github.com/geodynamics/aspect/blob/master/unit_tests/parse_map_to_double_array.cc#L199
